### PR TITLE
Refine OAuth flow

### PR DESCRIPTION
**Pull request by annando**
The process in requesting an access token has been revised according to the documention by Mastodon and this article: https://aaronparecki.com/oauth-2-simplified/

Also several apps had been tested and their test results are noted.

### CODE REVIEW

**Comment by AlfredSK** [comment]
Btw. is there already a way in Friendica to show existing access tokens and to revoke access tokens? Or is this planned?
There is that page "connected apps" in settings which has never been used as far as I know. :-) I guess this would be the place to list the apps with access and to provide the possibility to revoke the app access tokens.
What do you think?

**Comment by annando** [comment]
Currently there is no way looking at the access tokens or to revoke them. But yeah, that place sounds right for this.

**Comment by tobiasd** [comment]
On the "connected apps" page there is a button to remove apps.

grafik

And the trash icon works as expected :)

File: src/Model/Item.php [hunk 1]
```diff
@@ -2914,11 +2914,11 @@ private static function addLinkAttachment(array $attachments, string $body, stri
 				$data['description'] = '';
 			}
 
-			if (!empty($data['author_name']) && !empty($data['provider_name'])) {
+			if (($data['author_name'] ?? '') == ($data['provider_name'] ?? '')) {
 				$data['author_name'] = '';
 			}
 
-			if (!empty($data['author_url']) && !empty($data['provider_url'])) {
+			if (($data['author_url'] ?? '') && ($data['provider_url'] ?? '')) {
```
Comment (MrPetovan) [inline src/Model/Item.php]: What is the purpose of this change? Did you mean to use a `==` like on line 2917?

**Comment by annando** [comment]
@tobiasd that part refers to the legacy OAuth implementation that uses completely different tables. But we can reuse and repurpose some stuff from it.

**Review by MrPetovan — CHANGES_REQUESTED** [review]

File: src/Model/Item.php (hunk 1 above)
Comment (annando) [reply src/Model/Item.php]: A typical late night/middle in the night/early morning coding.

File: database.sql [hunk 2]
```diff
@@ -1500,6 +1507,27 @@ CREATE TABLE IF NOT EXISTS `workerqueue` (
...
+	`application`.`client_id` AS `client_id`,
+	`application`.`client_secret` AS `client_secret`,
+	`application-token`.`code` AS `code`,
+	`application-token`.`access_token` AS `access_token`,
+	`application-token`.`created_at` AS `created_at`,
+	`application-token`.`scopes` AS `scopes`,
+	`application-token`.`read` AS `read`,
+	`application-token`.`write` AS `write`,
+	`application-token`.`follow` AS `follow`
+	FROM `application-token` INNER JOIN `application` ON `application-token`.`application-id` = `application`.`id`;
```
Comment (MrPetovan) [inline database.sql]:
```suggestion
	FROM `application-token`
	INNER JOIN `application` ON `application-token`.`application-id` = `application`.`id`;
```

**Review by MrPetovan — CHANGES_REQUESTED** [review]

**Review by MrPetovan — APPROVED** [review]

**Comment by tobiasd** [comment]
Why would you introduce a second place to manage apps connected with an account? ;-)

**Comment by MrPetovan** [comment]
Because they don't have the same authentication mechanism I would assume. I understand it doesn't make sense for end users though.

**Comment by annando** [comment]
In this PR I repurposed the existing (old) OAuth settings to work with the new OAuth. I assume that no one had worked with the old OAuth mechanism at all.

**Comment by tobiasd** [comment]
I did some trails a while back (they worked back then) but I disabled my trials again as it dit not really fit into my workflow.
